src/compiler/checker.ts

```diff
@@ -9884,12 +9884,8 @@ namespace ts {
             if (!strictNullChecks) {
                 return getSupertypeOrUnion(types);
             }
-            const primaryTypes = filter(types, t => !(t.flags & TypeFlags.Nullable));
-            if (!primaryTypes.length) {
-                return getUnionType(types, /*subtypeReduction*/ true);
-            }
-            const supertype = getSupertypeOrUnion(primaryTypes);
-            return supertype && getNullableType(supertype, getFalsyFlagsOfTypes(types) & TypeFlags.Nullable);
+            const supertype = getSupertypeOrUnion(map(types, getNonNullableType));
+            return supertype && getNullableType(supertype, getFalsyFlagsOfTypes(types) & (TypeFlags.Nullable | TypeFlags.Void));

```

Comment:
Here's what caused the addition `TypeFlags.Void` to `getFalsyFlagsOfTypes`. When `types` contains `void`, the old `filter(types ...` code did not remove it because `void` only has `TypeFlags.Void`. But `map(types, getNonNullableType)` relies on `getTypeFacts`, which treats `void` and `undefined` exactly the same. So it removes `void` and `undefined` both. This causes the following bug:

```ts
declare function f<T>(cb: () => T | undefined): T;
f(() => { });
  ~~~~~~~~~
  Argument of type '() => void' is not assignable to parameter of type '() => undefined'.
```

I would like getNonNullableType not to remove void, but I'm not sure how to do that without changing getTypeFacts, which I am not at **all** comfortable doing.